friendz with Android Audio API

BlackMinou opened the pull request:
friendz now uses the Android Audio API 
+ fixed the new_global_ref functions in Assets and Resources

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,25 @@ redef class Game
 	# ResizeButton
 	var button_size = new ResizeButton(self)
 
+	init
```
pourquoi avoir déplacé le init ?

BlackMinou replied:
Heu ben je trouvais qu'un init se trouvait mieux en première position

privat replied:
OK

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -1587,6 +1583,7 @@ redef class App
 	do
 		super
 		game = new Game
+		game.load_sounds(self)
```
pourquoi ne pas appeler load_sounds dans le init de game?

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,25 @@ redef class Game
+	do
+		music = app.load_sound("music.ogg").as(not null)
+		snd_click = app.load_sound("click.wav").as(not null)
+                snd_win = app.load_sound("level.wav").as(not null)
```
indentation louche

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,25 @@ redef class Game
+                snd_win = app.load_sound("level.wav").as(not null)
+                snd_duh = app.load_sound("duh.wav").as(not null)
+                snd_bing = app.load_sound("bing.wav").as(not null)
+                snd_whip = app.load_sound("whip.wav").as(not null)
```
le tas de as not null fait pas beau :(

BlackMinou replied:
C'est parce que l'API Audio Android a été faite avant l'abstraction, et que le load_sound de l'abstraction renvoie des nullable Sound, Je ferais les modifications nécessaire dans l'API Android dans une prochaine PR

privat replied:
En general on finit les API avant de bricoler les clients, mais ok...

BlackMinou replied:
Ouai c'est que je me suis rendu compte du problème en modifiant le client justement ^^

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -920,28 +904,29 @@ redef class Game
 
 	# Click sound
-	var snd_click = new Audio("click.wav")
+	var snd_click: Sound is writable
```
personnellement je pense que c'est une regression : avant, le nom du ficher ressource était à coté de l'attribut le représentant.

BlackMinou replied:
Je vois ce que tu veux dire mais dans ce cas ce serait un problème au niveau de l'abstraction ? Comme on load un son à partir de l'App en lui donnant le path, ça suggèrerait de la modifier pour qu'elle prenne un son à qui on a déjà donné le path. Dans ce cas on passe à une initialisation du son en 2 étapes en le créant d'abord avant de le charger. Il y a aussi d'autres possibilités, quel comportement te conviendrait le mieux ?

privat replied:
Aucune idée.

`app` est global donc on pourait charger directement en causant à `app`.

~~~nit
redef class Game
   var x: Sound = app.load_sound("x.wav")
end
~~~

ou passer par un service de Game pour eviter de causer directement à `app`

~~~nit
redef class Game
   var x: Sound = load_sound("x.wav")
   fun load_sound(filename: String): Sound
   do
       do return app.load_sound(finename).as(not null)
   end
end
~~~

Une autre idée à considerer (par pour cette PR) est le chargement en deux temps. C'est peut-etre une bonne approche : 1. on cree un objet qui identifie la resource ; 2. on charge la resource plus tard

~~~nit
var x = app.get_audio("x.wav")
assert x.is_loaded == false
x.load # charge le son
assert x.is_loaded == true
~~~

Le chargement pourait etre fait automatiquement lors de demande de service

~~~nit
var x = app.get_audio("x.wav")
assert x.is_loaded == false
x.play # charge implicitement puis joue
assert x.is_loaded == true
~~~

`app` devrait aussi pouvoir connaitres ses ressources, donc on pourait lui demander de charger des trucs à un moment donné pour éviter d'avoir des lags du au paresseux

~~~nit
var x = app.get_audio("x.wav")
assert x.is_loaded == false
app.load_all
assert x.is_loaded == true
x.play # direct play
~~~

xymus replied:
La couche supplémentaire serait intéressante. En plus on pourrait y glisser la gestion des erreurs! Et éventuellement juste noop les appels à play si le son n'est pas disponible.

Le son pourrait être chargé à l'instanciation lorsque possible, sinon être ajouté à une liste qui serait parcourue par un appel de méthode plus tard.

xymus commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -21,11 +22,20 @@ redef class App
+			return false
+		else
+			return super
+		end
```
This is uselessly hackish. The previous def deal with KeyEvents, but it does return true for all keys. Just update onKeyDown to return true only on intercepted events and it will fix your volume keys problem.

xymus commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -35,3 +45,14 @@ redef class AndroidPointerEvent
+		super
+                # Audio initialization
+                music = app.load_music("music.ogg")
+                app.default_mediaplayer.looping = true
```
Weird indentation...

xymus commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -21,11 +22,20 @@ redef class App
 	redef fun window_created
 	do
 		super
-
```
Did you have any reason to remove the last two empty lines? They looked like they were appropriate.

Morriar commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,25 @@ redef class Game
+		title
+	end
+
+	#loading sounds, must be redef by specific platforms
```
Is this supposed to be a synopsys? It's not looking like a synopsys.

Add a space after the `#` and change the first letter for an uppercase.

Also, maybe review the comment to explain that all specific sounds for the game are loaded here.

Morriar commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -13,7 +13,8 @@ module friendz_android
 import android::landscape
-
+import android::audio
+import android::assets_and_resources
```
Add a new blank line after the imports block.

privat commented:
@BlackMinou si tu veux etre dans la niteration, grouille-toi de finir la PR :)

BlackMinou commented:
Ping reroll !
Added error gestion in the sound API
there is now 3 differents ways of getting android sounds : 
* loading it from assets with `load_music` or `load_sound`
* from resources with `load_sound_from_res` or `load_music_from_res`
* the easy way, insranciate a `new Sound(name)` or a  `new Music(name)` then call `load` on it.
In every cases, if the sound or the music loading fails, it has an error and the operations on it are no-oped, so there is no runtime errors.
Modified friendz and mnit_simple to use this upgrade of the API.

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -920,28 +904,29 @@ redef class Game
 
 	# The background music resource. */
-	var music = new Audio("music.ogg")
+	var music: Music = new Music("music.ogg") is writable
```
useless static type and why writable?

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,26 @@ redef class Game
+	end
+
+	#loading sounds, must be redef by specific platforms
+	fun load_sounds(app: App)
```
app is useless

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,26 @@ redef class Game
 	# ResizeButton
 	var button_size = new ResizeButton(self)
 
+	init(app: App)
```
app is useless

privat commented on contrib/friendz/src/friendz.nit:
```diff
@@ -972,6 +957,26 @@ redef class Game
+		snd_win.load
+		snd_duh.load
+		snd_bing.load
+		snd_whip.load
```
can't the loading be done lazily at the first play?

privat commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -35,3 +45,12 @@ redef class AndroidPointerEvent
+	do
+		super
+                app.default_mediaplayer.looping = true
+		if not self.music_muted then music.play
```
I do not understand why I have something specific to do on android that cannot be done in app.nit

Morriar commented on lib/android/audio.nit:
```diff
@@ -101,13 +102,21 @@ private extern class NativeMediaPlayer in "Java" `{ android.media.MediaPlayer `}
+			return null;
+		}
+
+
```
:)

Morriar commented on lib/android/audio.nit:
```diff
@@ -145,9 +156,33 @@ private extern class NativeSoundPool in "Java" `{ android.media.SoundPool `}
+			int id =  self.load(afd, (int)priority);
+			return id;
+		}catch(Exception e){
+			return -1;
```
Shouldn't you keep that message to display it from the Nit side?

Maybe the sameway that is used in [native_mongodb](https://github.com/privat/nit/blob/master/lib/mongodb/native_mongodb.nit#L106-L111)?

@xymus what is your opinion on that?

Morriar commented on lib/android/audio.nit:
```diff
@@ -196,21 +235,49 @@ class SoundPool
+	private fun load_asset_fd(afd: NativeAssetFileDescriptor): Sound do
+		var resval = nsoundpool.load_asset_fd(afd, self.priority)
+		if  resval == -1 then
+			self.error = new Error("Unable to load sound from assets")
```
Yeah, here it can be usefull to reuse the exception message.

xymus commented on contrib/friendz/src/friendz_android.nit:
```diff
@@ -35,3 +45,12 @@ redef class AndroidPointerEvent
+	redef fun load_sounds(app: App)
+	do
+		super
+                app.default_mediaplayer.looping = true
```
Weird indent.

privat commented:
+1

R4PaSs commented:
:shipit:
